persist/fs/write.go

```diff
@@ -164,7 +170,17 @@ func (w *writer) Write(key string, data []byte) error {
 	return w.WriteAll(key, [][]byte{data})
 }
 
-func (w *writer) WriteAll(key string, data [][]byte) error {
+func (w *writer) WriteAll(key string, data [][]byte) (retErr error) {
+	if w.err != nil {
+		return w.err
+	}
+
+	defer func() {
+		if retErr != nil {
```

Comment:
actually `retErr` will get assigned when the function returns.

Compared against the disk I/O overhead, honestly I wouldn't worry too much about the defer overhead. The alternative is adding `w.err = err` in 10 other different places in this function, which seems quite messy.